### PR TITLE
add mau_trial_days config param.

### PR DESCRIPTION
**Pull request by ara4n**
only consider users MAU after they've been around N days.
This is an alternative implementation to https://github.com/matrix-org/synapse/pull/3739
as suggested by @neilisfragile, which is much simpler as you just hold off adding
users to the MAU table until they've been active for more than N days.

### CODE REVIEW

File: synapse/storage/monthly_active_users.py [hunk 1]
```diff
@@ -177,19 +177,36 @@ def user_last_seen_monthly_active(self, user_id):
...
+                (timestamp, created) = (rows[0][0], rows[0][1])
+                if not created:
+                    created = 0
+                trial = (timestamp - created * 1000) < mau_trial_ms
+                return (timestamp, trial)
+
+        return self.runInteraction(
+            "user_last_seen_monthly_active",
+            _user_last_seen_monthly_active
+        )
```
Comment (erikjohnston) [inline synapse/storage/monthly_active_users.py]: We're going to need to invalidated the cache when the users get promoted from trial users

File: synapse/storage/monthly_active_users.py [hunk 2]
```diff
@@ -177,19 +177,36 @@ def user_last_seen_monthly_active(self, user_id):
...
+            txn.execute(sql, (user_id,))
+            rows = txn.fetchall()
+            if not rows or not rows[0][0]:
+                return None
+            else:
+                (timestamp, created) = (rows[0][0], rows[0][1])
+                if not created:
+                    created = 0
+                trial = (timestamp - created * 1000) < mau_trial_ms
+                return (timestamp, trial)
```
Comment (erikjohnston) [inline synapse/storage/monthly_active_users.py]: Hang on, trial users aren't added to the mau table and so they'll never be picked up by the select here?

Comment (ara4n) [reply synapse/storage/monthly_active_users.py]: oh, crap, yes; my bad. should just be looking at the agr of the user in that instance and ignoring the MAU table entirely, which simplifies things even more and gets rid of the fugly left join

Comment (ara4n) [reply synapse/storage/monthly_active_users.py]: oh, crap, yes; my bad. should just be looking at the agr of the user in that instance and ignoring the MAU table entirely, which simplifies things even more and gets rid of the fugly left join

**Comment by erikjohnston** [comment]
Well I completely messed that up, so I tried to do it again: https://github.com/matrix-org/synapse/pull/3749

**Comment by ara4n** [comment]
obsoleted by #3749 - thanks @erikjohnston for picking it up.

**Pull request closed without merging**